### PR TITLE
Even more bom ut

### PR DESCRIPTION
**Pull request by lewisrenfrew**
- mainly just a big refactor of the BomChangeService (which github has handily hidden since the diff is so big - make sure you expand that one out if you are reviewing to see what I actually changed)
 - the general idea was to split out the messy spaghetti of the main 'ProcessTreeUpdate' method  into hopefully more understandable smaller methods. One for traversing, one for processing each sub assembly, and then specialised ones for the different kinds of updates we do to bom tree nodes (add, delete, undo delete, replace)
 - I also removed the added complexity of keeping the tree (in memory) up to date and returning that to the client. Now the state of the tree in the db is the single source of truth, and I query the database after any updates commit to guarantee the most-up-to-date tree is returned to the client - trade off is I have to hit the db again obviosuly to rebuild the tree, which makes the PostBomTree request a bit slower.
 - all kinds of bug fixes

### CODE REVIEW

File: src/Service.Host/client/src/components/BomUtility/BomUtility.js [hunk 1]
```diff
@@ -119,14 +119,19 @@ function BomUtility() {
...
             crNumber && contextMenu === null
                 ? {
                       mouseX: e.clientX + 2,
                       mouseY: e.clientY - 6,
                       part: target.innerText,
-                      canReplace: Number(crNumber) !== detail.addChangeDocumentNumber,
+                      canDelete: !detail.deleteReplaceSeq,
+                      canReplace:
+                          Number(crNumber) !== detail.addChangeDocumentNumber &&
+                          !detail.addReplaceSeq &&
```
Comment (andrewfraser73) [inline src/Service.Host/client/src/components/BomUtility/BomUtility.js:132]: does the !addReplaceSeq prevent chaining of change requests i.e. replace a component on one change request then change that change to another part number in another one.  not common but sometimes Jacki might do crazy stuff like that (worth asking)

**Review by andrewfraser73 — APPROVED** [review]
seems ok one tiny query about chaining of changes

File: src/Service.Host/client/src/components/BomUtility/BomUtility.js (hunk 1 above)
Comment (lewisrenfrew) [reply src/Service.Host/client/src/components/BomUtility/BomUtility.js:132]: Yeah - as it stands you can delete or replace a change that is involved in a deletion/replacement on a different change request

File: src/Domain.LinnApps/Boms/BomChangeService.cs [hunk 2]
```diff
@@ -345,13 +103,207 @@ public void DeleteAllFromBom(string bomName, int crfNumber, int changedBy)
...
+        {
+            // todo - is this correct? - if so add a unit test to enshrine it
+            if (node.DeleteChangeDocumentNumber.GetValueOrDefault() != changeRequestNumber)
+            {
+                throw new InvalidBomChangeException(
+                    $"cannot undo delete of {node.Name} - it was deleted on a different change request.");
+            }
+
+            node.DeleteChangeDocumentNumber = null;
+            node.ChangeState = "LIVE"; // todo - check this is right - add to test if so
```
Comment (lewisrenfrew) [inline src/Domain.LinnApps/Boms/BomChangeService.cs:253]: @andrewfraser73 any thoughts on this todo? My thinking is if you undo a delete node then that node needs to go back to being 'LIVE' - can you think of any situation where that won't be the case?

**Pull request merged**